pennylane/interfaces/jax.py

```diff
@@ -373,3 +374,201 @@ def wrapped_exec_bwd(params, g):
+    """Execute a batch of tapes with JAX parameters on a device.
+
+    Args:
+        tapes (Sequence[.QuantumTape]): batch of tapes to execute
+        execute_fn (callable): The execution function used to execute the tapes
+            during the forward pass. This function must return a tuple ``(results, jacobians)``.
+            If ``jacobians`` is an empty list, then ``gradient_fn`` is used to
+            compute the gradients during the backwards pass.
+        gradient_kwargs (dict): dictionary of keyword arguments to pass when
+            determining the gradients of tapes
+        gradient_fn (callable): the gradient function to use to compute quantum gradients
+        _n (int): a positive integer used to track nesting of derivatives, for example
+            if the nth-order derivative is requested.
+        max_diff (int): If ``gradient_fn`` is a gradient transform, this option specifies
+            the maximum order of derivatives to support. Increasing this value allows
+            for higher order derivatives to be extracted, at the cost of additional
+            (classical) computational overhead during the backwards pass.
+
+    Returns:
+        list[list[float]]: A nested list of tape results. Each element in
+        the returned list corresponds in order to the provided tapes.
+    """
+    # Set the trainable parameters
+    for tape in tapes:
+        params = tape.get_parameters(trainable_only=False)
+        tape.trainable_params = qml.math.get_trainable_indices(params)
+
+    parameters = tuple(list(t.get_parameters()) for t in tapes)
+
+    if gradient_fn is None:
+        # PennyLane forward execution
+        return _execute_fwd_new(
+            parameters,
+            tapes=tapes,
+            execute_fn=execute_fn,
+            gradient_kwargs=gradient_kwargs,
+            _n=_n,
+        )
+
+    # PennyLane backward execution
+    return _execute_bwd_new(
+        parameters,
+        tapes=tapes,
+        execute_fn=execute_fn,
+        gradient_fn=gradient_fn,
+        gradient_kwargs=gradient_kwargs,
+        _n=_n,
+        max_diff=max_diff,
+    )
+
+
+def _execute_fwd_new(
+    params,
+    tapes,
+    execute_fn,
+    gradient_kwargs,
+    _n=1,
+):
+    """The auxiliary execute function for cases when the user requested
+    jacobians to be computed in forward mode (e.g. adjoint) or when no gradient function was
+    provided. This function is does not allow multiple derivatives."""
+
+    # pylint: disable=unused-variable
+    @jax.custom_jvp
+    def execute_wrapper(params):
+        new_tapes = [_copy_tape(t, a) for t, a in zip(tapes, params)]
+
+        with qml.tape.Unwrap(*new_tapes):
+            res, jacs = execute_fn(new_tapes, **gradient_kwargs)
+
+        res = _to_jax(res)
+
+        return res, jacs
+
+    @execute_wrapper.defjvp
+    def execute_wrapper_jvp(primal, tangents):
+        res, jacs = execute_wrapper(primal[0])
```

Comment:
Based on the signature of `execute_wrapper`, I'd guess that `primal[0]` are the tape parameters? What are the other elements in `primal`?

It would be nice to have a comment here for developers not familiar with the terminology (such as myself).

Reply:
primal only contains the parameters normally, but I cannot figure out why it is a list of list. See the comment above about the index [0]. I will add some comments.